### ISSUE TITLE
Add a new modifier for working with HLS (m3u8) files

### ISSUE THREAD

**Issue by ameshkov** [AdguardTeam/CoreLibs#1434]
We need a way to modify HLS playlists so that we could remove ad segments from them.

Rules examples: https://github.com/AdguardTeam/AdguardFilters/blob/master/EnglishFilter/sections/replace.txt
RFC: https://tools.ietf.org/html/rfc8216

- [x] Syntax spec
- [x] Implementation
- [ ] Update Knowledge base
- [ ] Port to tsurlfilter

**Comment by ngorskikh** [comment]
The example rules mostly aim to simply remove a segment with a matching URL along with all of its tags.
However, in some cases there's more complex logic, for example:
1. `/(#EXTINF:.*\n.*tvessaiprod\.nbcuni\.com\/video\/[\s\S]*?#EXT-X-DISCONTINUITY|#EXTINF:.*\n.*nbcuadops-vh.akamaihd.net\/i\/ads\/.*\n)//` here it seems the removal of a segment is conditional on whether it is followed by an X-DISCONTINUITY tag, and the following X-DISCONTINUITY tag is removed as well.
2. `/(#EXT-X-TARGETDURATION:[\S]*?)\n[\s\S]*?#EXT-X-DISCONTINUITY/\$1/` a segment is removed conditional on it being immediately preceded by one tag and immediately followed by another, and the following tag is removed.
3. `/#UPLYNK-SEGMENT: \S*\\,ad\s[\s\S]+?((#UPLYNK-SEGMENT: \S+\\,segment)|(#EXT-X-ENDLIST))/\$1/` in this case it seems that anything following a line matching one pattern is removed, until a line matching another pattern is encountered.

I propose two syntaxes:
1. One for simply removing a segment by URL pattern, something like `$removesegment=/<regex>/d`, which will automatically remove the segment's tags, with `d` being the optional flag meaning "only match if the next segment has the X-DISCONTINUITY tag, and remove the X-DISCONTINUITY flag from the next segment".
2. Another for removing everything between two lines matching the "start" and "stop" condition, something like `$removesegment=/<start regex>/<stop regex>/be` where `be` is the optional flags, with `b` meaning "also remove the start line", and `e` meaning "also remove the stop line".

**Comment by ameshkov** [comment]
The syntax you propose seems too close to the original `$replace` rules.
Can we maybe come up with something closer to the HLS spec?

A playlist may be represented as an object with some specific properties (a list of segments for example).

There are several HLS parsers you can take a look at. Here's [a very popular JS parser](https://www.npmjs.com/package/m3u8-parser) and there's a chance that it is used on all those websites.

It would be interesting to analyze our rules result using that parser. What exact parts of the playlists are getting removed? Is it some segments? Or maybe some custom tags?

**Comment by ngorskikh** [comment]
Parsing the HLS playlist and removing matching segments/tags would be easy enough. The issue is that there are some rules which are not that trivial, examples:
1. ```||ssaiplayback.*.boltdns.net/playback/once/v*/hls/*.m3u8$replace=/(#EXT-X-TARGETDURATION:[\S]*?)\n[\s\S]*?#EXT-X-DISCONTINUITY/\$1/,domain=news.com.au``` – a whole chunk of the playlist after some matching line and up to and including a line matching another pattern is removed.
2. ```||manifest.*.theplatform.com/*/*.m3u8?*referer=www.nbc.com$replace=/(#EXTINF:.*\n.*tvessaiprod\.nbcuni\.com\/video\/[\s\S]*?#EXT-X-DISCONTINUITY|#EXTINF:.*\n.*nbcuadops-vh.akamaihd.net\/i\/ads\/.*\n)//,domain=player.theplatform.com,important``` – some segment (along with it's tag) is removed only if it's followed by a discontinuity tag, and that tag is also removed.
3. ```||uplynk.com/*.m3u8$replace=/#UPLYNK-SEGMENT: \S*\\,ad\s[\s\S]+?((#UPLYNK-SEGMENT: \S+\\,segment)|(#EXT-X-ENDLIST))/\$1/``` again, everything starting from a line matching some condition is yanked out, up until a line matching another condition. I couldn't find the documentation for this non-standard `#UPLYNK-SEGMENT` "tag".

By the way, in cases 1 and 2 (and probably also 3) the rule might break playback because it blindly removes a discontinuity tag (it's inserted when e.g. the codec changes between segments, and the player might continue to try to decode the following tags using the wrong codec). It seems that in these particular cases the rules rely on another discontinuity tag being present before the removed section, but that doesn't sound very reliable.

Maybe we should ask the filter developers which capabilities exactly do they need to efficiently manipulate HLS playlists? :)

Presumably, the minimum functionality that is required is:
a. Remove a segment if its URL matches the pattern. All of the segment's tags are also removed. A discontinuity tag may be inserted before the next segment if it's required.
Probably we also need to:
b. Remove a segment if some of its tags match the pattern (this might be what example 3 is actually trying to achieve).
c. Do whatever examples 1 and 2 are supposed to do :)

**Comment by ameshkov** [comment]
@ngorskikh filter developers are not aware of the complexity of HLS, these rules are the result of their experiments, they are not necessarily valid.

Can we maybe check if the desired result can be achieved if we use proper HLS semantics instead of removing random tags.

You will probably need the help of @AdguardTeam/filters-maintainers with providing examples of the ads these rules actually block.

**Comment by ngorskikh** [comment]
@ameshkov After speaking with @AdamWr and @Alex-302, I think I now have a better understanding of what the filters are trying to do, thanks to examples provided by Adam.

Summary of ad blocking transformations seen so far:
1. "Remove all segments with a matching URL".
2. "Remove all segments that have a matching tag". There are some tags in wide use that are not defined by the RFC. Some of them are [documented](https://docs.aws.amazon.com/mediatailor/latest/ug/hls-ad-markers.html), others, such as `#UPLYNK-SEGMENT`, are not, but we can probably deduce their meaning from example playlists. We'll have to support these non-standard tags as well.
3. "Remove all segments until the first segment with a discontinuity tag" (probably based on the assumption that the first segments in the playlist are ads, and they are in a different format from the main content). We could provide a feature like "remove all segments before the first segment with a matching tag/URL" to support this.

I propose a syntax like `$hls=/<regex>/<flags>`. Flags control the behaviour as explained above. This can later be expanded to `$hls=/<regex>/<replacement>/<flags>` if we ever need to replace a segment instead of removing it. The rule should do nothing if the first bytes of the response are not `#EXTM3U`, i.e. if the response is not in fact an HLS playlist. In terms of interaction with other types of rules, it should behave just as if it was a `$replace` rule.

**Comment by ameshkov** [comment]
1. We'd better take a look at the real examples of number 3, maybe there's a better solution?
2. Can we avoid regular expressions or at least not make it the default option? Regexes are rather complicated and for matching URLs people are used to the basic URL matching syntax already.

**Comment by AdamWr** [comment]
Regarding:
```
||manifest.*.theplatform.com/*/*.m3u8?*referer=www.nbc.com$replace=/(#EXTINF:.*\n.*tvessaiprod\.nbcuni\.com\/video\/[\s\S]*?#EXT-X-DISCONTINUITY|#EXTINF:.*\n.*nbcuadops-vh.akamaihd.net\/i\/ads\/.*\n)//,domain=player.theplatform.com,important
```
This rule can be checked for example here - [https://www.nbc.com/late-night-with-seth-meyers/video/bidens-hot-mic-moment-prince-andrews-teddy-bear-collection-late-nights-news-of-the-week/959877711](https://adguardteam.github.io/AnonymousRedirect/redirect.html?url=https://www.nbc.com/late-night-with-seth-meyers/video/bidens-hot-mic-moment-prince-andrews-teddy-bear-collection-late-nights-news-of-the-week/959877711)

As far as I understand, it removes "segments" with video ads - https://regex101.com/r/Kup9IL/1

<hr>

```
||uplynk.com/*.m3u8$replace=/#UPLYNK-SEGMENT: \S*\\,ad\s[\s\S]+?((#UPLYNK-SEGMENT: \S+\\,segment)|(#EXT-X-ENDLIST))/\$1/
```

This one can be checked here - [https://www.fox.com/watch/1df49c86f927f5c840e8856c335e188b/](https://adguardteam.github.io/AnonymousRedirect/redirect.html?url=https://www.fox.com/watch/1df49c86f927f5c840e8856c335e188b/)
For example - https://regex101.com/r/ogqRZQ/1

In both cases US IP is required.

**Comment by AdamWr** [comment]
And maybe one more example.


Rules:
```
/playback/v3/videoPlaybackInfo$replace=/("forecastTimeline" : \[)[\s\S]*?(]\,[\s\S]*?"vendorAttributes")/\$1\$2/,domain=go.discovery.com|discoveryplus.com|go.tlc.com|watch.hgtv.com|investigationdiscovery.com|sciencechannel.com
/playback/v3/videoPlaybackInfo$replace=/("vendorAttributes" : {)[\s\S]*?}\,[\s\S]*?("streaming")/\$1}}\,\n\$2/,domain=go.discovery.com|discoveryplus.com|go.tlc.com|watch.hgtv.com|investigationdiscovery.com|sciencechannel.com
||uplynk.com/preplay2/*.mpd$replace=/<Period duration="[a-zA-Z\.\d]+" id="[a-zA-Z\-\d]+"><AssetIdentifier schemeIdUri="urn:com:uplynk:ad-asset-id".*?<\/AdaptationSet><\/Period>//,domain=go.discovery.com|discoveryplus.com|go.tlc.com|watch.hgtv.com|investigationdiscovery.com|sciencechannel.com
```

Link - [https://go.discovery.com/video/alaska-the-last-frontier-discovery/a-whole-new-frontier](https://adguardteam.github.io/AnonymousRedirect/redirect.html?url=https://go.discovery.com/video/alaska-the-last-frontier-discovery/a-whole-new-frontier)

It's also uplynk, but rather new and instead of `.m3u8` there is `.mpd` file format.
Segments with ads contains:
```
AssetIdentifier schemeIdUri="urn:com:uplynk:ad-asset-id"
```

For `.mpd`, it removes video ads segments:
https://regex101.com/r/cQKXEq/1

For "videoPlaybackInfo":
https://regex101.com/r/BFYHD3/1
https://regex101.com/r/YN64LX/1
It's required to remove ad breaks.
As far as I understand, it's JSON and as I recall there is an another task for that https://github.com/AdguardTeam/CoreLibs/issues/1447

**Comment by ngorskikh** [comment]
> 1. We'd better take a look at the real examples of number 3, maybe there's a better solution?
> 2. Can we avoid regular expressions or at least not make it the default option? Regexes are rather complicated and for matching URLs people are used to the basic URL matching syntax already.

@ameshkov 
1. Can't find any examples, the link in the comment above the rule leads to a video where there are no ads, and this is the only such rule in our filters. Maybe it's not very relevant :)
2. I suppose URL filter syntax can be allowed to remove segments by URL, but some characters, like `/`, `,` and `$`, will have to be escaped: `$hls=\/ab\$c\/*\,ad.ts,important`.

**Comment by ameshkov** [comment]
2. We also use "string or regexp" parameters a lot in scriptlets. It means that in a normal case we just check if the specified value is a part of the string, but if the value starts and ends with `/`, we evaluate it as a regular expression.

Regarding the rule syntax, is there any place where we need to remove individual tags and not the whole segments?
How should we behave when there are multiple rules matching the same request?

**Comment by ngorskikh** [comment]
2. We need to escape `$` and `,` inside the modifier value to not throw off the modifier parsing (`$` separates modifiers from the URL pattern and `,` separates them from each other), and `/` I think should be escaped as well because a URL pattern may start and end with a `/`, in which case it would be confused with a regex.

Removing individual tags is not something that our existing rules do, and in any case it can be done with `$replace` rules just as easily, so would be redundant.

Multiple rules matching the same request should all be appiled at once (meaning the response should be parsed only once). Any segments matching any of the matched rules should be removed. Exception rules should be handled exactly as for `$replace` rules.

**Comment by ameshkov** [comment]
2. Naturally, escape rules are the same for all modifiers. What's your opinion on the general concept of "string or regexp"?:

    ```
    $hls=substring
    $hls=/regexp/
    ```

When all the questions are sorted out, we need to have a final spec, use this as a template:
https://kb.adguard.com/en/general/how-to-create-your-own-ad-filters#removeheader-modifier

**Comment by ngorskikh** [comment]
2. Sorry, I missed the point about substrings, not URL patterns, in your previous post :harold: Well, substring matching is naturally a subset of URL pattern matching, so it is automatically supported. It's not very useful, though, as all existing rules that match segment URLs use a more complicated pattern than just "contains substring", so I think it's still useful to allow full URL patterns (with the only limitation being that some characters have to be escaped), e.g.:
```
$hls=substring                                                             // remove segments with URLs containing "substring"
$hls=||redirector.googlevideo.com\/videoplayback^?*&source=dclk_video_ads  // remove segments with URLs matching the pattern
$hls=||seamless.mtvnservices.com\/media^?beaconUrl=*.ts&isAd=true          // remove segments with URLs matching the pattern
$hls=/^#UPLYNK-SEGMENT:.*\,ad\$/t                                          // remove segments with a tag that matches the regex
```

**Comment by ameshkov** [comment]
@ngorskikh do we match only URLs or do we match the full tag + value?

**Comment by ngorskikh** [comment]
@ameshkov I think we should match only URLs unless specifically instructed to match tags: 
1. `$hls=examplesubstring`, `$hls=example.org\/video\/*.ts` and `$hls=/example\.org\/video\/.*\.ts/` should remove segments based on the URL. 
2. `$hls=/#UPLYNK-SEGMENT:.*\,ad/t` (note the `t` option) should remove segments which have a matching tag (and ignore the URL).

I don't see the value in one rule matching both tags and URLs: they are completely different entities: URL are, well, URLs, and tags are just opaque strings with (almost) completely arbitrary format. (The RFC says that they must start with `#EXT`, however, as can be seen in the previous posts, there are non-standard tags that don't. However, they all start with `#` since anything not starting with `#` is interpreted as a URL according to the RFC).

Of course, when a segment is removed, we should remove all tags that belonged to that segment, otherwise the playlist will become corrupted. In case of a tag that applies to multiple segments, we should remove it when all of the segments that it applies to are removed.

**Comment by ngorskikh** [comment]
@ameshkov How about this for a final spec:

### `$hls` modifier

`$hls` rules modify the response of a matching request. They are intended as a convenient way to remove segments from [HLS playlists (RFC 8216)](https://datatracker.ietf.org/doc/html/rfc8216). Note that the word "segment" in this document means either a "Media Segment" or a "playlist" as part of a "Master Playlist": `$hls` rules do not distinguish between a "Master Playlist" and a "Media Playlist".

### Syntax
* `||example.org^$hls=urlpattern` – remove segments whose URL matches the URL pattern `urlpattern`.  The pattern works just like the one in basic URL rules, however, the characters `/`, `$` and `,` must be escaped with `\` inside `urlpattern`.
* `||example.org^$hls=/regex/options` – remove segments where the URL, or, if a certain option is present, one of the tags, matches the regular expression `regex`. `options` modify the behaviour of the rule as described below.

Available `options` are:
* `t` – instead of testing the segment's URL, test each of the segment's tags against the regex. A segment with a matching tag is removed.
* `i` – make the regular expression case-insensitive.

The characters `/`, `$` and `,` must be escaped with `\` inside `regex`.

### Exceptions

Basic URL exceptions shall not disable `$hls` rules. They can be disabled as described below:
* `@@||example.org^$hls` – disable all `$hls` rules for responses from URLs matching `||example.org^`.
* `@@||example.org^$hls=text` – disable all `$hls` rules with the value of the `hls` modifier equal to `text` for responses from URLs matching `||example.org^`.
* `$hls` rules can also be disabled by `$document`, `$content` and `$urlblock` exception rules.

### Restrictions
* `$hls` rules are only allowed in trusted filters.
* `$hls` rules are not compatible with any other modifiers except `$domain`, `$third-party`, `$app`, `$important`, `$match-case`, and `$xmlhttprequest`.
* `$hls` rules only apply to HLS playlists, which are UTF-8 encoded text starting with the line `#EXTM3U`. Any other response will not be modified by these rules.

### Notes
When multiple `$hls` rules match the same request, their effect is cumulative.

### Examples
* `||example.org^$hls=\/videoplayback^?*&source=dclk_video_ads` — removes all segments with the matching URL.
* `||example.org^$hls=/\/videoplayback\/?\?.*\&source=dclk_video_ads/i` — achieves more or less the same with
  a regex instead of a URL pattern.
* `||example.org^$hls=/#UPLYNK-SEGMENT:.*\,ad/t` — removes all segments which have the matching tag.

**Comment by ngorskikh** [comment]
@ameshkov Previous comment updated.

**Comment by ameshkov** [comment]
@ngorskikh so a segment always starts with `EXTINF` and ends with `EXT-X-DISCONTINUITY`, right?

**Comment by ngorskikh** [comment]
@ameshkov No, there is no such requirement. There are some tags that are required according to the RFC, e.g. `#EXTINF` is required for all media segments (it specifies the length of the segment). Other tags are optional. The order of the tags is irrelevant as long as it doesn't change to which segments they apply to.

For example, `#EXT-X-DISCONTINUITY` is used to signal that timestamps (or something) may "jump" in the following segment. Because apparently the decoder may break on such discontinuity, so the player needs to know about it in advance to re-initialize the decoder before feeding it the segment (or something). This tag may be used when there's an ad insert in the middle of the main content, but it is not strictly related to ads.
